bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java

```diff
@@ -433,6 +445,109 @@ public void readComplete(int rc, LedgerHandle lh,
+                            } else if (toSend instanceof ByteBufList) {
+                                updateAverageEntrySize(((ByteBufList) toSend).readableBytes());
+                            }
+                        }
+                        AtomicInteger numCompleted = new AtomicInteger(0);
+                        AtomicBoolean completed = new AtomicBoolean(false);
+
+                        WriteCallback multiWriteCallback = new WriteCallback() {
+                            @Override
+                            public void writeComplete(int rc, long ledgerId, long entryId, BookieId addr, Object ctx) {
+                                if (rc != BKException.Code.OK) {
+                                    LOG.error("BK error writing entry for ledgerId: {}, entryId: {}, bookie: {}",
+                                            ledgerId, entryId, addr, BKException.create(rc));
+                                    if (completed.compareAndSet(false, true)) {
+                                        ledgerFragmentMcb.processResult(rc, null, null);
+                                    }
+                                } else {
+                                    numEntriesWritten.inc();
+                                    if (ctx instanceof Long) {
+                                        numBytesWritten.registerSuccessfulValue((Long) ctx);
+                                    }
+                                    if (LOG.isDebugEnabled()) {
+                                        LOG.debug("Success writing ledger id {}, entry id {} to a new bookie {}!",
+                                                ledgerId, entryId, addr);
+                                    }
+                                    if (numCompleted.incrementAndGet() == newBookies.size()
+                                            && completed.compareAndSet(false, true)) {
+                                        ledgerFragmentMcb.processResult(rc, null, null);
+                                    }
+                                }
+                            }
+                        };
+
+                        for (BookieId newBookie : newBookies) {
+                            long startWriteEntryTime = MathUtils.nowInNano();
+                            bkc.getBookieClient().addEntry(newBookie, lh.getId(),
+                                    lh.getLedgerKey(), entry.getEntryId(), toSend,
+                                    multiWriteCallback, dataLength, BookieProtocol.FLAG_RECOVERY_ADD,
+                                    false, WriteFlag.NONE);
+                            writeDataLatency.registerSuccessfulEvent(
+                                    MathUtils.elapsedNanos(startWriteEntryTime), TimeUnit.NANOSECONDS);
+                        }
+                        toSend.release();
+                    }
+                    if (lastEntryId == endEntryId) {
```

Comment:
When `lastEntryId == endEntryId`, we needn't callback `ledgerFragmentMcb.processResult(BKException.Code.OK, null, null);`.

Reply:
Good point! Updated